### DISCUSSION TITLE
Livewire Variables Getting Mixed Up

### DISCUSSION THREAD

**Discussion by bobwurtz** [livewire/livewire#3913] [1. Help]
Hey All,

I'm stuck on an issue and could use some advice and/or guidance. I use Livewire to manage a payment processing page in my application. I use Stripe as my credit card processor. In `mount`, I call a function that creates the payment intent and saves the payment intent ID to the user (`user` is a public variable tied to a Model).

```
public function setStripeInfo()
{
        // payment intent
        Stripe::setApiKey(config('services.stripe.sk'));
        $paymentIntent = PaymentIntent::create([
            'amount' => 50,
            'currency' => 'usd',
            'payment_method_types' => ['card', 'card_present'],
            'capture_method' => 'manual',
            'on_behalf_of' => $this->stripe_act,
            'transfer_data' => [
                'destination' => $this->stripe_act,
            ]
        ]);

        $this->user->payment_intent_id = $paymentIntent->id;
        $this->user->save();

        $this->clientSecret = $paymentIntent->client_secret;
}
```

Later, after a customer has added items to their cart, I attempt to capture the payment intent.

```
$pi = $stripe->paymentIntents->retrieve(
            $this->user->payment_intent_id,
            []
        );

        $total = round($pi->amount / 100, 2);

        $stripe_fee = round( 0.05 + ( $total * 0.027 ), 2);

try
        {
            $payment = $stripe->paymentIntents->capture(
                $pi->id,
                [
                    'transfer_data' =>
                    [
                        'amount' => $pi->amount - ( $stripe_fee * 100 ),
                    ]
                ]
            );
        }
        catch(\Stripe\Exception\InvalidRequestException $e)
        {
            \Log::error($e,
                [
                    'use_case_id' => $this->useCaseId,
                    'organization_id' => $this->organization->id,
                    'organization' => $this->organization->name,
                    'user_type' => $this->user_type,
                    'user_id' => $this->user_id,
                    'invalid_payment_intent' => $pi->id,
                    'payment_intent_id_check' => $this->user->payment_intent_id,
                ]
            );

            $this->paymentError = true;
            return true;
        }
```

I am experiencing some strange behavior on roughly 1% of capture requests - `$this->user->payment_intent_id` appears to be pulling the wrong payment intent ID. I have been back and forth with Stripe several times now and I am fairly certain the issue is not on their end. Based on what I can gather, my application is attempting to capture a different payment intent than the one that was declared in `mount`. Previously, I was using a `public $paymentIntentId` variable and using that to capture the payment. Switching to storing the payment intent in a table has helped cut down on the errors but they still occur.

I have a few questions:

1. Does Livewire have any known issues of confusing the state of one page with the state of a different active page? This payment application runs simultaneously in multiple locations. I deploy my application using Vapor, if that matters.

2. Is there a better way for me to store the payment intent ID as the user moves through the checkout process? What is the best practice there?

Thank you!

**Comment by denizozturk** [comment]
If you're using associated livewire components, it's definitely best to write the data to the session first, then define it to the variable inside the component. In the mount functions of the components, you can define the value from the session to the variables. I have always used session in such transactions.

**Reply by bobwurtz** [reply]
Thanks so much for the response. When you say 'write the data to the session first' what do you mean by that? Can you provide an example?

**Reply by denizozturk** [reply]
I'm not sure I fully understand the problem, but I think you're saying that an ID value that you saved to the user model is giving an error. I save this type of ID or temporary information like `session()->put('paymentId', 123456)` and use it as `session()->get('paymentId')` in transactions. In the background, you can save the value you keep in the session to the database.

**Reply by bobwurtz** [reply]
Thank you - this is very interesting. I never would have thought to save data like that.

To provide more context, here are the steps this Livewire component takes:
1. Initialize and display the page (the guest and paymentId are created in mount)
2. Add items to cart
3. Accept card payment and confirm with Stripe API (this is where the problem occurs)
4. Create new guest and new paymentId for the next customer (page data clears but page does not refresh)

My issue is that Livewire seems to be returning the wrong paymentId on a small number of payment attempts (roughly 10 out of every 500 requests - 2%). I don't know why this is happening or why it would only happen on a small percentage of requests. I will switch to using a session value to store the paymentId and see if this fixes the problem.
